subprojects/language-groovy/src/main/java/org/gradle/api/tasks/compile/GroovyCompile.java

```diff
@@ -135,11 +203,166 @@ private void warnIfCompileAvoidanceEnabled() {
             Compiler<GroovyJavaJointCompileSpec> delegatingCompiler = groovyCompilerFactory.newCompiler(spec);
-            compiler = new CleaningGroovyCompiler(delegatingCompiler, getOutputs());
+            CleaningGroovyCompiler cleaningGroovyCompiler = new CleaningGroovyCompiler(delegatingCompiler, getOutputs());
+            IncrementalCompilerFactory<GroovyJavaJointCompileSpec> factory = new IncrementalCompilerFactory<GroovyJavaJointCompileSpec>(
+                getFileOperations(), getStreamHasher(), getCompileCaches(), getBuildOperationExecutor(), getStringInterner(), getFileSystemSnapshotter(), getFileHasher()
+            );
+            if (getOptions().isIncremental()) {
+                return factory.makeIncremental(cleaningGroovyCompiler, getPath(), new IncrementalInputsAdapter(inputChanges, getStableSources()), getStableSources().getAsFileTree(), ".groovy", new Function<CompilationSourceDirs, SourceToNameConverter>() {
+                    @Override
+                    public SourceToNameConverter apply(CompilationSourceDirs sourceDirs) {
+                        return new GroovySourceToNameConverter(sourceDirs, sourceClassesMapping);
+                    }
+                });
+            } else {
+                compiler = cleaningGroovyCompiler;
+            }
         }
         return compiler;
     }
 
+    /**
+     * @since 5.6
+     */
+    @Incubating
+    @Incremental
+    @PathSensitive(PathSensitivity.RELATIVE) // Java source files are supported, too. Therefore we should care about the relative path.
+    @InputFiles
+    protected FileCollection getStableSources() {
+        return stableSources;
+    }
+
+    /**
+     * @since 5.6
+     */
+    @Inject
+    protected FileSystemSnapshotter getFileSystemSnapshotter() {
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * @since 5.6
+     */
+    @Inject
+    protected BuildOperationExecutor getBuildOperationExecutor() {
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * @since 5.6
+     */
+    @Inject
+    protected FileOperations getFileOperations() {
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * @since 5.6
+     */
+    @Inject
+    protected StreamHasher getStreamHasher() {
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * @since 5.6
+     */
+    @Inject
+    protected FileHasher getFileHasher() {
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * @since 5.6
+     */
+    @Inject
+    protected StringInterner getStringInterner() {
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * @since 5.6
+     */
+    @Inject
+    protected GeneralCompileCaches getCompileCaches() {
+        throw new UnsupportedOperationException();
+    }
+
+    private static class IncrementalInputsAdapter implements IncrementalTaskInputs {
```

Comment:
This should work the other way round: `IncrementalInputs` should be adapted to input changes, so they can be queried for changes to the sources and the classpath separately. Then most of the code in `RecompilationSpecProvider` shouldn't be necessary any more if `InputChanges` are used, since it is possible to query for each kind of change separately. Note that the alignment which is done as part of `processClasspathChanges` is also not necessary, since the changes are now reported in a way which does not need to be mapped again.

Reply:
I didn't adapt `IncrementalInputs` or `IncrementalTaskInputs` - I make `RecompilationSpecProvider` polymorphic instead.